airflow/upgrade/rules/undefined_jinja_varaibles.py

```diff
@@ -0,0 +1,121 @@
+        """Replicates the logic in BaseOperator.render_template() to
+        cover all the cases needed to be checked.
+        """
+        if isinstance(rendered_content, six.string_types):
+            return set(re.findall(r"{{(.*?)}}", rendered_content))
+
+        elif isinstance(rendered_content, (tuple, list, set)):
+            debug_error_messages = set()
+            for element in rendered_content:
+                debug_error_messages.union(self._check_rendered_content(element))
+            return debug_error_messages
+
+        elif isinstance(rendered_content, dict):
+            debug_error_messages = set()
+            for key, value in rendered_content.items():
+                debug_error_messages.union(self._check_rendered_content(str(value)))
+            return debug_error_messages
+
+    def _render_task_content(self, task, content, context):
+        completed_rendering = False
+        errors_while_rendering = []
+        while not completed_rendering:
+            # Catch errors such as {{ object.element }} where
+            # object is not defined
+            try:
+                renderend_content = task.render_template(content, context)
+                completed_rendering = True
+            except Exception as e:
+                undefined_variable = re.sub(" is undefined", "", str(e))
+                undefined_variable = re.sub("'", "", undefined_variable)
+                context[undefined_variable] = dict()
+                message = "Could not find the object '{}'".format(undefined_variable)
+                errors_while_rendering.append(message)
+        return renderend_content, errors_while_rendering
+
+    def _task_level_(self, task):
+        messages = {}
+        task_instance = TaskInstance(task=task, execution_date=timezone.utcnow())
+        context = task_instance.get_template_context()
+        for attr_name in task.template_fields:
+            content = getattr(task, attr_name)
+            if content:
+                rendered_content, errors_while_rendering = self._render_task_content(
+                    task, content, context
+                )
+                debug_error_messages = list(
+                    self._check_rendered_content(rendered_content)
+                )
+                messages[attr_name] = errors_while_rendering + debug_error_messages
+
+        return messages
+
+    def _dag_level_(self, dag):
+        dag.template_undefined = jinja2.DebugUndefined
+        tasks = dag.tasks
+        messages = {}
+        for task in tasks:
+            error_messages = self._task_level_(task)
+            messages[task.task_id] = error_messages
+        return messages
+
+    def check(self, dagbag=None):
+        if not dagbag:
+            dag_folder = conf.get("core", "dags_folder")
+            dagbag = DagBag(dag_folder)
+        dags = dagbag.dags
+        messages = []
+        for dag_id, dag in dags.items():
+            dag_messages = self._dag_level_(dag)
```

Comment:
Do you think it is worth omitting checking some DAGs in some cases?

Reply:
Didn't exactly get this?
Which DAGs could be omitted from checking?

Reply:
DAGs in which the user has decided about a specific behavior should not be checked. 

```python
with DAG(dag_id="case_a"):
```
```python
with DAG(dag_id="case_b", template_undefined=jinja2.StrictUndefined):
```
```python
with DAG(dag_id="case_c", template_undefined=jinja2.DebugUndefined):
```
```python
with DAG(dag_id="case_d", template_undefined=jinja2.Undefined):
```
Only `dag_a` should be checked. The others should be omitted as they will work properly in Airlfow 2.0.

Reply:
Any progress?

Reply:
I missed this comment - my fault.
How exactly would we check this?

My thought process - 
This would require the check to read the Python scripts in which the DAGs have been defined. Then read line by line where the DAG is defined and then check whether `template_undefined` is defined or not.

Reply:
A bit more thinking on this. We could - 
1. Regex search the python files for a pattern where `template_undefined=jinja2.` exists.
2. If a match occurs - search `dag_id="` using regex around the match and get the dag_id
3. The gathered dag_id's then can be skipped for a check. 

This approach does omit cases such as - 
```
dag_name = "example_dag"
example = DAG(dag_id=dag_name, template_undefined=jinja2.Undefined)
```


How would you suggest writing this?

Reply:
It seems to me that it is only possible if we make changes to the DAG constructor and get_template_env method.
https://github.com/apache/airflow/blob/d8aa24b729de4e80cd2d71f99d04cbabf5f3b36c/airflow/models/dag.py#L238
```diff
-        template_undefined: Type[jinja2.StrictUndefined] = jinja2.StrictUndefined,
+        template_undefined: Optional[Type[jinja2.StrictUndefined]] = None,
```
https://github.com/apache/airflow/blob/d8aa24b729de4e80cd2d71f99d04cbabf5f3b36c/airflow/models/dag.py#L968
```diff
-            'undefined': self.template_undefined,
+            'undefined': self.template_undefined or jinja2.StrictUndefined,,

```
After making the above changes, we can check the class attribute to test our condition without any major problems.

Reply:
I see. Makes sense.
I believe this would require a new PR for the change?

Should I raise one for the same?

Reply:
Why does this require new PR? I think you can update this PR.

Reply:
Got it. Working on the changes.

Reply:
Added the changes w.r.t to `v1-10-stable` as the target branch.
The code changes you had suggested were w.r.t to `master` branch.

Also added the logic to skip check for DAG when `template_undefined` is explicitly mentioned